### PR TITLE
Add optimizers (AdaMax, AdaDelta, RMSprop) and ordering optimizer classes

### PR DESCRIPTION
**Pull request by Jyun1998**
Before submitting this PR, check the [contribution guidelines](CONTRIBUTING.md).

Make sure your code is formatted: `pre-commit run --all-files`.

---

- ran pre-commit
- ran test_optimizers.py

---

```
<class 'optimizers.AdaDelta'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-0.00447209, -0.00447209, -0.00447209, ..., -0.00447209, -0.00447209, -0.00447209], dtype=float32), array([-0.00447209], dtype=float32)], 'second': array([-0.00447209], dtype=float32)}
<class 'optimizers.Adamax'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-0.01, -0.01, -0.01, ..., -0.01, -0.01, -0.01], dtype=float32), array([-0.01], dtype=float32)], 'second': array([-0.01], dtype=float32)}
<class 'optimizers.Adagrad'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-0.1, -0.1, -0.1, ..., -0.1, -0.1, -0.1], dtype=float32), array([-0.1], dtype=float32)], 'second': array([-0.1], dtype=float32)}
<class 'optimizers.Adam'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-0.1, -0.1, -0.1, ..., -0.1, -0.1, -0.1], dtype=float32), array([-0.1], dtype=float32)], 'second': array([-0.1], dtype=float32)}
<class 'optimizers.AdamW'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-0.1, -0.1, -0.1, ..., -0.1, -0.1, -0.1], dtype=float32), array([-0.1], dtype=float32)], 'second': array([-0.1], dtype=float32)}
<class 'optimizers.RMSprop'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-1, -1, -1, ..., -1, -1, -1], dtype=float32), array([-1], dtype=float32)], 'second': array([-1], dtype=float32)}
<class 'optimizers.SGD'>
{'first': [array([1, 1, 1, ..., 1, 1, 1], dtype=float32), array([1], dtype=float32)], 'second': array([1], dtype=float32)} {'first': [array([-0.1, -0.1, -0.1, ..., -0.1, -0.1, -0.1], dtype=float32), array([-0.1], dtype=float32)], 'second': array([-0.1], dtype=float32)}
..
----------------------------------------------------------------------
Ran 1 test in 0.077s

OK
```

### CODE REVIEW

File: python/mlx/optimizers.py [hunk 1]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+
+    .. math::
+
+        v_{t+1} &= \beta v_t + (1 - \beta) g_t^2 \\
+        w_{t+1} &= w_t - \lambda \frac{g_t}{\sqrt{v_{t+1} + \epsilon}}
+
+    [1]: Tieleman, T. and Hinton, G. 2012. Lecture 6.5-rmsprop, coursera: Neural networks for machine learning
+    """
+
+    def __init__(self, learning_rate: float, beta: float = 0.99, eps: float = 1e-8):
```
Comment (awni) [inline python/mlx/optimizers.py:154]: Any reason you went with `beta` instead of `alphas` as in PyTorch? If not, let's just use `alpha` for the sake of consistency?

File: python/mlx/optimizers.py [hunk 2]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+        """Performs the Adagrad parameter update and stores :math:`v` in the
+        optimizer state."""
+        lr = self.learning_rate
+        eps = self.eps
+
+        v = state.get("v", mx.zeros_like(gradient))
+        v = v + mx.square(gradient)
+        state["v"] = v
+
+        return parameter - lr * gradient / (mx.sqrt(v) + eps)
```
Comment (awni) [inline python/mlx/optimizers.py:226]: This is a bit different than the stated implementation which has `eps` inside the `\sqrt`

File: python/mlx/optimizers.py [hunk 3]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+
+class Adagrad(Optimizer):
+    r"""Implementation of the Adagrad optimizer [1].
+
+    Our Adagrad implementation follows the original paper. In detail,
+
+    .. math::
+
+        v_{t+1} &= v_t + g_t^2 \\
+        w_{t+1} &= w_t - \lambda \frac{g_t}{\sqrt{v_{t+1} + \epsilon}}
```
Comment (awni) [inline python/mlx/optimizers.py:193]: I think the `epsilon` should be outside the `\sqrt`

File: python/mlx/optimizers.py [hunk 4]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+    def __init__(self, learning_rate: float, eps: float = 1e-8):
+        super().__init__()
+
+        self.learning_rate = learning_rate
+        self.eps = eps
+
+        if self.learning_rate < 0.0:
+            raise ValueError(
+                f"Adagrad learning rate should be >=0, {self.learning_rate} was provided instead"
+            )
```
Comment (awni) [inline python/mlx/optimizers.py:208]: Do we need this constraint? I see no reason not to allow negative LRs..

File: python/mlx/optimizers.py [hunk 5]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+
+        v_{t+1} &= \rho v_t + (1 - \rho) g_t^2 \\
+        \Delta w_{t+1} &= - \frac{\sqrt{s_t + \epsilon}}{\sqrt{v_{t+1} + \epsilon}} g_t \\
+        s_{t+1} &= \rho s_t + (1 - \rho) \Delta w_{t+1}^2 \\
+        w_{t+1} &= w_t + \Delta w_{t+1}
+
+    [1]: Zeiler, M.D., 2012. ADADELTA: an adaptive learning rate method. arXiv preprint arXiv:1212.5701.
+    """
+
+    def __init__(self, learning_rate: float, rho: float = 0.95, eps: float = 1e-6):
```
Comment (awni) [inline python/mlx/optimizers.py:244]: Should we default `rho=0.9` as in `torch.optim`?

File: python/mlx/optimizers.py [hunk 6]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+    r"""Implementation of the AdaDelta optimizer with learning rate[1].
+
+    Our AdaDelta implementation follows the original paper. In detail,
+
+    .. math::
+
+        v_{t+1} &= \rho v_t + (1 - \rho) g_t^2 \\
+        \Delta w_{t+1} &= - \frac{\sqrt{s_t + \epsilon}}{\sqrt{v_{t+1} + \epsilon}} g_t \\
+        s_{t+1} &= \rho s_t + (1 - \rho) \Delta w_{t+1}^2 \\
+        w_{t+1} &= w_t + \Delta w_{t+1}
```
Comment (awni) [inline python/mlx/optimizers.py:239]: This is missing the learning rate right?

File: python/mlx/optimizers.py [hunk 7]
```diff
@@ -140,6 +140,144 @@ def apply_single(
...
+        s = state.get("s", mx.zeros_like(gradient))
+
+        v = rho * v + (1 - rho) * mx.square(gradient)
+        d = -mx.sqrt(s + eps) / mx.sqrt(v + eps) * gradient
+        s = rho * s + (1 - rho) * mx.square(d)
+
+        state["v"] = v
+        state["s"] = s
+
+        return parameter + d
```
Comment (awni) [inline python/mlx/optimizers.py:278]: I would change this slightly to be not negate `d` since it is squared in the next line.

Then do `parameter - lr * d`.

Also you are missing the `lr`.

Comment (awni) [reply python/mlx/optimizers.py:278]: And maybe update the doc string accordingly.

File: python/mlx/optimizers.py (hunk 3 above)
Comment (awni) [inline python/mlx/optimizers.py:193]: Please include the `Args` section (as in the SGD example) and you can mention that `\lambda` is the learning rate here.

Same comment for the other optimizers.

**Review by awni — CHANGES_REQUESTED** [review]
This is awesome! A few changes and then we can merge it!

File: python/mlx/optimizers.py (hunk 3 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py:193]: I guess pre-existing code like adam, adamw also didn't have explanation aboute the \lambda.

In torch and tensorflow optimizers code, they usually do not have detailed explanation, so I did like it.

Do you want to include Args section for all other pre-existing functions as well?

File: python/mlx/optimizers.py (hunk 2 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py:226]: This is pre-existing codes. I fixed the docstring equation where eps should be outside the sqrt

File: python/mlx/optimizers.py (hunk 1 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py]: Agreed and changed.

File: python/mlx/optimizers.py (hunk 4 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py]: deleted

File: python/mlx/optimizers.py (hunk 3 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py]: It was pre-existing code but I fixed :)

File: python/mlx/optimizers.py (hunk 2 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py:222]: It was pre-existing code but I fixed :)  According to pytorch doc and paper, eps should be outside the sqrt

File: python/mlx/optimizers.py (hunk 5 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py]: yes changed according to https://pytorch.org/docs/stable/generated/torch.optim.Adadelta.html

File: python/mlx/optimizers.py (hunk 6 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py]: I think my local change was not applied. Changed according to https://pytorch.org/docs/stable/generated/torch.optim.Adadelta.html

File: python/mlx/optimizers.py (hunk 7 above)
Comment (Jyun1998) [reply python/mlx/optimizers.py]: yes changed according to https://pytorch.org/docs/stable/generated/torch.optim.Adadelta.html

File: python/mlx/optimizers.py (hunk 3 above)
Comment (awni) [reply python/mlx/optimizers.py]: I won't ask you to add it for all of them but at least moving forward for the new ones. If you are up for adding the Args section for others that would also be really great!

**Review by awni — CHANGES_REQUESTED** [review]
This is looking fantastic. If you don't mind adding the Args sections for at least the new Optimizers then we can land this after that!

**Comment by Jyun1998** [comment]
> This is looking fantastic. If you don't mind adding the Args sections for at least the new Optimizers then we can land this after that!

Adding to existing optimizers are not hard, so I will add all for the integrations :)

**Comment by awni** [comment]
@Jyun1998 still planning to update this?

**Comment by Jyun1998** [comment]
> @Jyun1998 still planning to update this?

Yes, I was busy doing interviews haha. just added and please check it!

**Review by awni — APPROVED** [review]
🚀 Merging this! Thanks a ton for the changes!
